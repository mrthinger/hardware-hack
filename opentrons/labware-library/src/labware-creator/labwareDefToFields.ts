import { getUniqueWellProperties } from '@opentrons/shared-data'
import type {
  Coordinates,
  LabwareDefinition2,
  LabwareWellGroup,
} from '@opentrons/shared-data'
import type { LabwareFields, BooleanString } from './fields'

// NOTE: this is just String() with some typing for flow
const boolToBoolString = (b: boolean): BooleanString => (b ? 'true' : 'false')

export function labwareDefToFields(
  def: LabwareDefinition2,
  adapterDefinitions: LabwareDefinition2[]
): LabwareFields | null {
  const allUniqueWellGroupProps = getUniqueWellProperties(def)

  if (allUniqueWellGroupProps.length === 0) {
    console.warn(
      'cannot get unique well props for definition, maybe it has missing/incomplete "groups" data?',
      def
    )
    return null
  }

  const {
    xSpacing,
    ySpacing,
    shape,
    depth,
    metadata,
    totalLiquidVolume,
    xOffsetFromLeft,
    yOffsetFromTop,
  } = allUniqueWellGroupProps[0]

  const homogeneousWells =
    allUniqueWellGroupProps.length === 1 &&
    shape !== null &&
    depth !== null &&
    totalLiquidVolume !== null

  // TODO: Ian 2019-08-28 once LC supports multiple groupds, num of rows / num of columns should
  // come from getUniqueWellProperties. Using `ordering` won't extend to multiple groups
  const gridRowsNum = def.ordering[0].length
  const gridColumnsNum = def.ordering.length

  const regularColumnSpacing = xSpacing !== null
  const regularRowSpacing = ySpacing !== null

  let labwareType: LabwareFields['labwareType'] | null = null

  if (
    def.metadata.displayCategory === 'wellPlate' ||
    def.metadata.displayCategory === 'tubeRack' ||
    def.metadata.displayCategory === 'aluminumBlock' ||
    def.metadata.displayCategory === 'reservoir' ||
    def.metadata.displayCategory === 'tipRack'
  ) {
    labwareType = def.metadata.displayCategory
  }

  if (
    !homogeneousWells ||
    !regularRowSpacing ||
    !regularColumnSpacing ||
    labwareType === null ||
    !(shape !== null && depth !== null && totalLiquidVolume !== null)
  ) {
    console.warn(
      'this labware def is more heterogeneous than LC can currently support',
      {
        homogeneousWells,
        regularColumnSpacing,
        regularRowSpacing,
        labwareType,
        allUniqueWellGroupProps,
      }
    )
    return null
  }

  const firstGroup: LabwareWellGroup | undefined = def.groups[0]
  const firstGroupBrand = firstGroup?.brand
  const zDimension = def.dimensions.zDimension
  const compatibleAdapters: Record<string, number> =
    def.stackingOffsetWithLabware != null
      ? Object.entries(
          def.stackingOffsetWithLabware as Record<string, Coordinates>
        ).reduce<Record<string, number>>((acc, [loadName, offset]) => {
          const adapterZDimension = Object.values(adapterDefinitions).find(
            def => def.parameters.loadName === loadName
          )?.dimensions.zDimension
          if (adapterZDimension != null) {
            acc[loadName] = adapterZDimension + zDimension - offset.z
          }

          return acc
        }, {})
      : {}
  const compatibleModules: Record<string, number> =
    def.stackingOffsetWithModule != null
      ? Object.entries(
          def.stackingOffsetWithModule as Record<string, Coordinates>
        ).reduce<Record<string, number>>((acc, [moduleModel, offset]) => {
          acc[moduleModel] = zDimension - offset.z
          return acc
        }, {})
      : {}

  return {
    // NOTE: Ian 2019-08-26 these LC-specific fields cannot easily/reliably be inferred
    tubeRackInsertLoadName: null,
    aluminumBlockType: null,
    aluminumBlockChildType: null,

    // We assume all tipracks are snug upon import
    handPlacedTipFit: labwareType === 'tipRack' ? 'snug' : null,

    labwareType,
    footprintXDimension: String(def.dimensions.xDimension),
    footprintYDimension: String(def.dimensions.yDimension),
    labwareZDimension: String(def.dimensions.zDimension),

    // Missing values or zeroes for these fields should be translated to null
    gridRows: gridRowsNum > 0 ? String(gridRowsNum) : null,
    gridColumns: gridColumnsNum > 0 ? String(gridColumnsNum) : null,
    gridSpacingX: xSpacing == null || xSpacing === 0 ? null : String(xSpacing),
    gridSpacingY: ySpacing == null || ySpacing === 0 ? null : String(ySpacing),

    gridOffsetX: String(xOffsetFromLeft),
    gridOffsetY: String(yOffsetFromTop),

    homogeneousWells: boolToBoolString(homogeneousWells),
    regularRowSpacing: boolToBoolString(regularRowSpacing),
    regularColumnSpacing: boolToBoolString(regularColumnSpacing),

    wellVolume: String(totalLiquidVolume),
    wellBottomShape: metadata.wellBottomShape ?? null,
    wellDepth: String(depth),
    wellShape: shape.shape,

    // used with circular well shape only
    wellDiameter: shape.shape === 'circular' ? String(shape.diameter) : null,

    // used with rectangular well shape only
    wellXDimension:
      shape.shape === 'rectangular' ? String(shape.xDimension) : null,
    wellYDimension:
      shape.shape === 'rectangular' ? String(shape.yDimension) : null,

    brand: def.brand.brand,
    brandId: def.brand.brandId != null ? def.brand.brandId.join(',') : null, // comma-separated values
    groupBrand: firstGroupBrand?.brand,
    groupBrandId: firstGroupBrand?.brandId?.join(',') ?? undefined,

    // NOTE: intentionally null these fields, do not import them
    loadName: null,
    displayName: null,
    compatibleAdapters,
    compatibleModules,
  }
}
