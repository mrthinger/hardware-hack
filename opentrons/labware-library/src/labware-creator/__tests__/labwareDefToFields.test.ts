import { vi, describe, it, expect } from 'vitest'
import { labwareDefToFields } from '../labwareDefToFields'
import {
  fixture_12_trough,
  fixture_24_tuberack,
  fixture_96_plate,
  fixture_irregular_example_1,
} from '@opentrons/shared-data/labware/fixtures/2'
import type { LabwareDefinition2 } from '@opentrons/shared-data'

vi.mock('../../definitions')

describe('labwareDefToFields', () => {
  it('fixture_96_plate', () => {
    const def = fixture_96_plate
    const result = labwareDefToFields(
      def as LabwareDefinition2,
      [def, def] as LabwareDefinition2[]
    )
    expect(result).toEqual({
      labwareType: 'wellPlate',
      tubeRackInsertLoadName: null,
      aluminumBlockType: null,
      aluminumBlockChildType: null,
      handPlacedTipFit: null,

      footprintXDimension: String(def.dimensions.xDimension),
      footprintYDimension: String(def.dimensions.yDimension),
      labwareZDimension: String(def.dimensions.zDimension),

      gridRows: '8',
      gridColumns: '12',
      gridSpacingX: '9',
      gridSpacingY: '9',
      gridOffsetX: String(def.wells.A1.x),
      gridOffsetY: '11.24',

      homogeneousWells: 'true',
      regularRowSpacing: 'true',
      regularColumnSpacing: 'true',

      wellVolume: '380',
      wellBottomShape: 'flat',
      wellDepth: '10.54',
      wellShape: 'circular',

      wellDiameter: '6.4',

      // used with rectangular well shape only
      wellXDimension: null,
      wellYDimension: null,

      brand: 'generic',
      brandId: null,

      loadName: null, // should be cleared
      displayName: null, // should be cleared

      compatibleAdapters: {},
      compatibleModules: {},
    })
  })

  it('fixture_12_trough', () => {
    // make sure rectangular wells + single row works as expected
    const def = fixture_12_trough
    const result = labwareDefToFields(
      def as LabwareDefinition2,
      [def, def] as LabwareDefinition2[]
    )

    expect(result?.labwareType).toEqual('reservoir')
    expect(result?.gridSpacingY).toBe(null) // single row -> null Y-spacing
    expect(result?.wellDiameter).toBe(null)
    expect(result).toMatchSnapshot()
  })

  it('fixture_irregular_example_1 should return null (until multi-grid labware is supported in LC)', () => {
    const def = fixture_irregular_example_1
    const result = labwareDefToFields(
      def as LabwareDefinition2,
      [def, def] as LabwareDefinition2[]
    )
    expect(result).toEqual(null)
  })

  it('fixture_24_tuberack should match snapshot', () => {
    const def = fixture_24_tuberack
    const result = labwareDefToFields(
      def as LabwareDefinition2,
      [def, def] as LabwareDefinition2[]
    )

    expect(result?.labwareType).toEqual('tubeRack')
    expect(result?.brand).toBe('Opentrons')
    expect(result?.brandId).toBe('649020')
    expect(result?.groupBrand).toBe('tube brand here')
    expect(result?.groupBrandId).toBe('tube123,other123')

    expect(result).toMatchSnapshot()
  })
})
