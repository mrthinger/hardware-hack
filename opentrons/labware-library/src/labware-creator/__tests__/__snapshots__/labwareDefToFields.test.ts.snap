// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`labwareDefToFields > fixture_12_trough 1`] = `
{
  "aluminumBlockChildType": null,
  "aluminumBlockType": null,
  "brand": "USA Scientific",
  "brandId": "1061-8150",
  "compatibleAdapters": {},
  "compatibleModules": {},
  "displayName": null,
  "footprintXDimension": "127.76",
  "footprintYDimension": "85.8",
  "gridColumns": "12",
  "gridOffsetX": "13.94",
  "gridOffsetY": "42.9",
  "gridRows": "1",
  "gridSpacingX": "9.09",
  "gridSpacingY": null,
  "groupBrand": undefined,
  "groupBrandId": undefined,
  "handPlacedTipFit": null,
  "homogeneousWells": "true",
  "labwareType": "reservoir",
  "labwareZDimension": "44.45",
  "loadName": null,
  "regularColumnSpacing": "true",
  "regularRowSpacing": "true",
  "tubeRackInsertLoadName": null,
  "wellBottomShape": "v",
  "wellDepth": "42.16",
  "wellDiameter": null,
  "wellShape": "rectangular",
  "wellVolume": "22000",
  "wellXDimension": "8.33",
  "wellYDimension": "71.88",
}
`;

exports[`labwareDefToFields > fixture_24_tuberack should match snapshot 1`] = `
{
  "aluminumBlockChildType": null,
  "aluminumBlockType": null,
  "brand": "Opentrons",
  "brandId": "649020",
  "compatibleAdapters": {},
  "compatibleModules": {},
  "displayName": null,
  "footprintXDimension": "127.75",
  "footprintYDimension": "85.5",
  "gridColumns": "6",
  "gridOffsetX": "18.21",
  "gridOffsetY": "10.07",
  "gridRows": "4",
  "gridSpacingX": "19.89",
  "gridSpacingY": "19.28",
  "groupBrand": "tube brand here",
  "groupBrandId": "tube123,other123",
  "handPlacedTipFit": null,
  "homogeneousWells": "true",
  "labwareType": "tubeRack",
  "labwareZDimension": "84",
  "loadName": null,
  "regularColumnSpacing": "true",
  "regularRowSpacing": "true",
  "tubeRackInsertLoadName": null,
  "wellBottomShape": "v",
  "wellDepth": "42",
  "wellDiameter": "8.5",
  "wellShape": "circular",
  "wellVolume": "2000",
  "wellXDimension": null,
  "wellYDimension": null,
}
`;
